File: benchmarks/csharp/MacroBench.cs
// SPDX-License-Identifier: MIT
// .NET Macrobenchmark Suite — equivalent to .JADE macro_bench.cpp

using System;
using System.Diagnostics;

class MacroBench
{
    static long N = 100_000_000;
    static int trials = 5;

    // 1. JSON scan
    static long JsonScan(long n)
    {
        long count = 0;
        for (long i = 0; i < n; i++)
        {
            long ch = i & 127;
            if (ch != 32 && ch != 9 && ch != 10)
                count++;
        }
        return count;
    }

    // 2. ECS update
    static long EcsUpdate(long n)
    {
        long sum = 0;
        for (long i = 0; i < n; i++)
            sum += i * 5 + 3;
        return sum;
    }

    // 3. Matmul inner
    static long MatmulInner(long n)
    {
        long sum = 0;
        for (long i = 0; i < n; i++)
            sum += i * i;
        return sum;
    }

    // 4. String interner
    static long StringIntern(long n)
    {
        long hash = 0;
        for (long i = 0; i < n; i++)
            hash = hash * 31 + i;
        return hash;
    }

    // 5. Expression evaluator
    static long ExprEval(long n)
    {
        long result = 0;
        for (long i = 0; i < n; i++)
        {
            if ((i & 1) == 1)
                result += i * 3;
            else
                result += i * 5;
        }
        return result;
    }

    static void Main(string[] args)
    {
        if (args.Length > 0) N = long.Parse(args[0]);
        if (args.Length > 1) trials = int.Parse(args[1]);

        Console.WriteLine("=== .NET Macrobenchmark Suite ===");
        Console.WriteLine($"Iterations: {N}, Trials: {trials}\n");

        var benches = new (string name, Func<long, long> fn)[]
        {
            ("json_scan", JsonScan),
            ("ecs_update", EcsUpdate),
            ("matmul_inner", MatmulInner),
            ("string_intern", StringIntern),
            ("expr_eval", ExprEval),
        };

        foreach (var (name, fn) in benches)
        {
            fn(N); // warmup
            var sw = Stopwatch.StartNew();
            for (int t = 0; t < trials; t++)
                fn(N);
            sw.Stop();
            double nsPer = sw.Elapsed.TotalMilliseconds / trials / N * 1e6;
            Console.WriteLine($"{name,-22} {nsPer,10:F3} ns/iter");
        }
    }
}

File: benchmarks/luajit/macro_bench.lua
#!/usr/bin/env luajit
-- Macrobenchmarks for LuaJIT — equivalent to .JADE macro_bench.cpp

local N = tonumber(arg[1]) or 100000000
local trials = tonumber(arg[2]) or 5

-- 1. JSON scan
local function json_scan(n)
    local count = 0
    for i = 0, n - 1 do
        local ch = i % 128
        if ch ~= 32 and ch ~= 9 and ch ~= 10 then
            count = count + 1
        end
    end
    return count
end

-- 2. ECS update
local function ecs_update(n)
    local sum = 0
    for i = 0, n - 1 do
        sum = sum + i * 5 + 3
    end
    return sum
end

-- 3. Matmul inner
local function matmul_inner(n)
    local sum = 0
    for i = 0, n - 1 do
        sum = sum + i * i
    end
    return sum
end

-- 4. String interner
local function string_intern(n)
    local hash = 0
    for i = 0, n - 1 do
        hash = hash * 31 + i
    end
    return hash
end

-- 5. Expression evaluator
local function expr_eval(n)
    local result = 0
    for i = 0, n - 1 do
        if i % 2 == 1 then
            result = result + i * 3
        else
            result = result + i * 5
        end
    end
    return result
end

local benches = {
    {"json_scan", json_scan},
    {"ecs_update", ecs_update},
    {"matmul_inner", matmul_inner},
    {"string_intern", string_intern},
    {"expr_eval", expr_eval},
}

print("=== LuaJIT Macrobenchmark Suite ===")
print(string.format("Iterations: %d, Trials: %d\n", N, trials))

for _, b in ipairs(benches) do
    local name, fn = b[1], b[2]
    fn(N) -- warmup
    local start = os.clock()
    for rep = 1, trials do
        fn(N)
    end
    local elapsed = os.clock()
    local ns_per = (elapsed / trials / N) * 1e9
    print(string.format("%-22s %10.3f ns/iter", name, ns_per))
end
